src/ray/gcs/gcs_server/gcs_server.cc

```diff
@@ -239,19 +243,35 @@ void GcsServer::InitGcsNodeManager(const GcsInitData &gcs_init_data) {
 
 void GcsServer::InitGcsHeartbeatManager(const GcsInitData &gcs_init_data) {
   RAY_CHECK(gcs_node_manager_);
-  gcs_heartbeat_manager_ = std::make_shared<GcsHeartbeatManager>(
-      heartbeat_manager_io_service_, /*on_node_death_callback=*/
-      [this](const NodeID &node_id) {
-        main_service_.post(
-            [this, node_id] { return gcs_node_manager_->OnNodeFailure(node_id); },
-            "GcsServer.NodeDeathCallback");
-      });
-  // Initialize by gcs tables data.
-  gcs_heartbeat_manager_->Initialize(gcs_init_data);
-  // Register service.
-  heartbeat_info_service_.reset(new rpc::HeartbeatInfoGrpcService(
-      heartbeat_manager_io_service_, *gcs_heartbeat_manager_));
-  rpc_server_.RegisterService(*heartbeat_info_service_);
+  auto node_death_callback = [this](const NodeID &node_id) {
+    main_service_.post(
+        [this, node_id] { return gcs_node_manager_->OnNodeFailure(node_id); },
+        "GcsServer.NodeDeathCallback");
+  };
+
+  if (RayConfig::instance().pull_based_healthcheck()) {
+    gcs_healthcheck_manager_ =
+        std::make_unique<GcsHealthCheckManager>(main_service_, node_death_callback);
```

Comment:
Why don't we run this in `heartbeat_manager_io_service_`?

Reply:
I'm trying to make it simple. As we know, it doesn't help a lot putting it on a separate thread for this case (we are doing this and seeing issues) since the bottleneck is not there.

I would prefer this way and if it's the bottleneck, tuning it later.

Reply:
Hmm still prefer not to rely on this sort of critical operation on the main thread as I've seen main thread being blocked > 10 seconds pretty often in current GCS

Reply:
But that's not the because of the health check right? Think about this, if it's overloaded, it'll check with lower frequency. My theory is that it's not going to make any difference and introducing a thread is not going to fix this issue.